addons/l10n_es_edi_facturae/models/l10n_es_edi_facturae_certificate.py

```diff
@@ -0,0 +1,121 @@
+    date_start = fields.Datetime(readonly=True, help="The date on which the certificate starts to be valid")
+    date_end = fields.Datetime(readonly=True, help="The date on which the certificate expires")
+    company_id = fields.Many2one(comodel_name='res.company', default=lambda self: self.env.company, required=True, readonly=True)
+    is_valid = fields.Boolean(string="is the certificate date valid", compute='_compute_is_valid', store=False, readonly=True)
+
+    @api.depends('date_start', 'date_end')
+    def _compute_is_valid(self):
+        for certificate in self:
+            certificate.is_valid = certificate.date_start < fields.Datetime.now() < certificate.date_end
+
+    @ormcache('self.content', 'self.password')
+    def _decode_certificate(self):
+        """
+        Return some certificate data concerning its validity
+        :return tuple: encoded certificate, private key, decrypted certificate
+        """
+        self.ensure_one()
+        _unused, certificate, *_unused = pkcs12.load_key_and_certificates(b64decode(self.content), self.password.encode())
+        return certificate.not_valid_before, certificate.not_valid_after, certificate.serial_number
+
+    # -------------------------------------------------------------------------
+    # LOW-LEVEL METHODS
+    # -------------------------------------------------------------------------
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        certificates = super().create(vals_list)
+        for certificate in certificates:
+            try:
+                cert_date_start, cert_date_end, serial_number = certificate._decode_certificate()
+            except ValueError:
+                raise UserError(_('There has been a problem with the certificate, some usual problems can be:\n'
+                                        '\t- The password given or the certificate are not valid.\n'
+                                        '\t- The certificate content is invalid.'))
+            if datetime.datetime.now() > cert_date_end:
+                raise UserError(_('The certificate is expired since %s') % certificate.date_end)
+            # Assign extracted values from the certificate
+            certificate.write({'serial_number': serial_number, 'date_start': cert_date_start, 'date_end': cert_date_end,})
+        return certificates
+
+    # -------------------------------------------------------------------------
+    # BUSINESS METHODS
+    # -------------------------------------------------------------------------
+    def sign_xml(self, edi_data, sig_data):
+        """
+        Signs the given XML data with the certificate and private key.
+        :param str edi_data: The XML data to sign.
+        :param dict sig_data: The signature data to use.
+        :return str: The signed XML data as a sting.
+        """
+        self.ensure_one()
+        if not self.is_valid:
```

Comment:
it's a public method and this line is the only ensuring the user has access to the certificate, maybe add a group to verify the user has the appropriate rights?